syslog-ng-rs/syslog-ng-common/tests/panicking-parser.rs

```diff
@@ -0,0 +1,192 @@
+    where C: FnOnce(),
+          P: FnOnce(pid_t)
+{
+    match unsafe { fork() } {
+        0 => {
+            child_callback();
+            Ok(())
+        }
+        x if x > 0 => {
+            parent_callback(x);
+            Ok(())
+        }
+        -1 | _ => Err(()),
+    }
+}
+
+fn assert_child_commits_suicide<C>(child_callback: C)
+    where C: FnOnce()
+{
+    let parent_callback = |child_pid| {
+        let mut status = 0;
+        let options = 0;
+        unsafe {
+            let result = waitpid(child_pid, &mut status, options);
+            assert!(result != -1);
+            assert!(WIFSIGNALED(status));
+            assert_eq!(SIGABRT, WTERMSIG(status));
+        };
+    };
+
+    fork_with_callbacks(child_callback, parent_callback).unwrap();
+}
+
+#[test]
+fn test_native_parser_proxy_new_wont_panic_even_if_the_proxy_panics() {
+    assert_child_commits_suicide(|| {
+        set_up_test();
+
+        let cfg = GlobalConfig::new(0x0308);
+        let _ = native_parser_proxy_new(cfg.raw_ptr());
+    });
+}
+
+#[test]
+fn test_native_parser_proxy_free_wont_panic_even_if_the_proxy_panics() {
+    assert_child_commits_suicide(|| {
+        set_up_test();
+
+        let proxy = ParserProxy::with_builder_and_parser(None, Some(PanickingParser(PhantomData)));
+        let _ = native_parser_proxy_free(Box::into_raw(Box::new(proxy)));
+    });
+}
+
+#[test]
+fn test_native_parser_proxy_set_option_wont_panic_even_if_the_proxy_panics() {
+    assert_child_commits_suicide(|| {
+        set_up_test();
+
+        let mut proxy =
+            ParserProxy::with_builder_and_parser(Some(PanickingParserBuilder(PhantomData)), None);
+        let key = CString::new("key").unwrap();
+        let value = CString::new("value").unwrap();
+        let _ = native_parser_proxy_set_option(&mut proxy, key.as_ptr(), value.as_ptr());
+    });
+}
+
+#[test]
+fn test_native_parser_proxy_init_wont_panic_even_if_the_proxy_panics() {
+    assert_child_commits_suicide(|| {
+        set_up_test();
+
+        let mut proxy =
+            ParserProxy::with_builder_and_parser(Some(PanickingParserBuilder(PhantomData)), None);
+        let _ = native_parser_proxy_init(&mut proxy);
+    });
+}
+
+#[test]
+fn test_native_parser_proxy_deinit_wont_panic_even_if_the_proxy_panics() {
+    assert_child_commits_suicide(|| {
+        set_up_test();
+
+        let mut proxy = ParserProxy::with_builder_and_parser(None,
+                                                             Some(PanickingParser(PhantomData)));
+        let _ = native_parser_proxy_deinit(&mut proxy);
+    });
+}
+
+#[test]
+fn test_native_parser_proxy_process_wont_panic_even_if_the_proxy_panics() {
+    assert_child_commits_suicide(|| {
+        set_up_test();
+
+        let mut proxy = ParserProxy::with_builder_and_parser(None,
+                                                             Some(PanickingParser(PhantomData)));
```

Comment:
Although the line `let mut proxy = ...` is the same length in each case, it is indented differently.